test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +510,438 @@ func TestAdoption(t *testing.T) {
+	// with Controller=true
+	ownerReference := metav1.OwnerReference{UID: uuid.NewUUID(), APIVersion: "apps/v1beta1", Kind: "StatefulSet", Name: rs.Name, Controller: &falseVar}
+	testPodControllerRefPatch(t, c, pod, &ownerReference, rs, 2)
+
+	// When the only OwnerReference of the pod points to the RS, but Controller=false
+	ownerReference = metav1.OwnerReference{UID: rs.UID, APIVersion: "extensions/v1beta1", Kind: "ReplicaSet", Name: rs.Name, Controller: &falseVar}
+	testPodControllerRefPatch(t, c, pod, &ownerReference, rs, 1)
+}
 
-	stopCh := make(chan struct{})
-	informers.Start(stopCh)
-	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	defer close(runControllerAndInformers(rm, informers))
 
-	// change the rs's selector to match both pods
-	patch := `{"metadata":{"labels":{"uniqueKey":"1"}}}`
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err := podClient.Patch(pod2.Name, types.StrategicMergePatchType, []byte(patch))
+	rs := newRS("rs", ns.Name, 2)
+	rs.Spec.MinReadySeconds = 5
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// First verify no pod is available
+	if rs.Status.AvailableReplicas != 0 {
+		t.Fatalf("Unexpected .Status.AvailableReplicas: Expected 0, saw %d", rs.Status.AvailableReplicas)
+	}
+
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap())
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %d, want 2", len(pods.Items))
+	}
+	podNum := len(pods.Items)
+	// Separate 2 pods into their own list
+	firstPodList := &v1.PodList{Items: pods.Items[:podNum-1]}
+	secondPodList := &v1.PodList{Items: pods.Items[1:]}
+	// Mark all pods running and ready
+	// First pod's LastTransitionTime is 5 seconds ago
+	markPodsRunningAndReady(t, c, firstPodList, ns.Name)
+	// This is made possible by sleeping for 5 seconds
+	time.Sleep(time.Second * 5)
+	// While second pod's LastTransitionTime is now
+	markPodsRunningAndReady(t, c, secondPodList, ns.Name)
+	waitPodsReady(t, c, int32(podNum), rs)
+
+	// Verify only one (second) pod is available
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
 	if err != nil {
-		t.Fatalf("Failed to patch pod2: %v", err)
+		t.Fatalf("Failed to obtain rs %s: %v", rs.Name, err)
 	}
-	t.Logf("patched pod2 = %#v", pod2)
-	// wait for the rs to select both pods and delete one of them
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 1)
-	}); err != nil {
-		t.Fatal(err)
+	// Polling is not needed as 5 seconds already passed
+	if newRS.Status.AvailableReplicas != 1 {
```

Comment:
It would be nice to test when Replicas, ReadyReplicas, and AvailableReplicas are all different from each other. You could do that by making 3 total Pods. Make one Running but not Ready, make one Running and Ready but not Available, make one Available.

Reply:
Sure! Let me add the sub-test.